### PR TITLE
CE-3551 Enable taps on Nice work! icon

### PR DESCRIPTION
**Pull request by kamilkoterba**
## Links

* http://wikia-inc.atlassian.net/browse/JIRA-3551

## Description

Enable adding nice work! by clicking on icon within recent-change component.
Component is used on recent-wiki-activity list and article-diff, so related controllers needed to be updated and common code for sending request moved to mixin.

## Reviewers

@Wikia/community-engineering, @Wikia/x-wing

### CODE REVIEW

File: front/main/app/mixins/revision-upvotes.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import getEditToken from '../utils/edit-token';
+
+export default Ember.Mixin.create({
+	/**
+	 * Sends request to MW API to upvote newId revision of title
+	 * @param {string} revisionId ID of revision that is upvoted
+	 * @param {string} title Text tilte of article in main namespace which revision is upvoted
+	 * @returns {Ember.RSVP.Promise}
+	 */
+	revisionUpvote(revisionId, title) {
```
Comment (adamkarminski) [inline front/main/app/mixins/revision-upvotes.js]: We shouldn't use mixins to communicate with APIs, it should a part of a model. @rogatty

Comment (kamilkoterba) [inline front/main/app/mixins/revision-upvotes.js]: Please note that this mixin it created to reuse common part in two models.

Comment (lukaszkonieczny) [inline front/main/app/mixins/revision-upvotes.js]: So it's sounds like a revision upvotes model or service

File: front/main/app/components/recent-change.js [hunk 2]
```diff
@@ -1,19 +1,71 @@
...
+				action: trackActions.error,
+				category: trackCategory,
+				label
+			});
+		},
+
+		actions: {
+			handleVote(revisionId, title) {
+				if (this.get('userUpvoted')) {
+					// TODO this.removeUpvote(this.get('currentUserUpvoteId'));
```
Comment (vforge) [inline front/main/app/components/recent-change.js]: `// @todo XXX`

File: front/main/app/components/recent-change.js [hunk 3]
```diff
@@ -1,19 +1,71 @@
...
+		trackSuccess(label) {
+			track({
+				action: trackActions.success,
+				category: trackCategory,
+				label
+			});
+		},
+
+		/**
+		 * Sends impression error tracking for recent-wiki-activity category
```
Comment (vforge) [inline front/main/app/components/recent-change.js]: This description seems obsolete - as function's name describe it pretty the same.

File: front/main/app/controllers/article-diff.js [hunk 4]
```diff
@@ -14,13 +14,13 @@ export default Ember.Controller.extend({
 		},
 
 		/**
-		 * Adds upvote for given revision
-		 *
-		 * @param {int} currentUserId
+		 * @param {string} revisionId Revision id that was upvoted
+		 * @param {string} title Text title from main namespace that revision was upvoted
+		 * @param {int} fromUser User id who upvoted
```
Comment (vforge) [inline front/main/app/controllers/article-diff.js]: It's `{number}`, not `{int}`

File: front/main/app/controllers/recent-wiki-activity.js [hunk 5]
```diff
@@ -4,5 +4,34 @@ export default Ember.Controller.extend({
 	application: Ember.inject.controller(),
 	siteHeadPinned: Ember.computed.oneWay('application.siteHeadPinned'),
 	queryParams: ['rc'],
-	rc: null
+	rc: null,
+
+	actions: {
+
+		/**
+		 * Adds upvote for given revision from provided user
```
Comment (vforge) [inline front/main/app/controllers/recent-wiki-activity.js]: Please do not add meaningless description for functions and function params - here and other places.

File: front/main/app/models/article-diff.js [hunk 6]
```diff
@@ -1,145 +1,135 @@
...
-								const errorMsg = resp.error.code === 'undofailure' ? 'main.undo-failure' : 'main.error';
+		/**
+		 * Sends request to MW API to undo newId revision of title
+		 * @param {*} summary Description of reason for undo to be stored as edit summary
+		 * CAUTION: if summary is {string} it will be used as a summary on MediaWiki side
+		 *          if summary is empty {Array} MediaWiki will provide default summary
+		 * @returns {Ember.RSVP.Promise}
+		 */
+		undo(summary = []) {
+			if (!summary) {
```
Comment (vforge) [inline front/main/app/models/article-diff.js]: This is not needed if you providing default value for the param.

File: front/main/app/models/article-diff.js [hunk 7]
```diff
@@ -1,145 +1,135 @@
...
-						dataType: 'json',
-						method: 'POST',
-						success: (resp) => {
-							if (resp && resp.edit && resp.edit.result === 'Success') {
-								resolve();
-							} else if (resp && resp.error) {
-								const errorMsg = resp.error.code === 'undofailure' ? 'main.undo-failure' : 'main.error';
+		/**
+		 * Sends request to MW API to undo newId revision of title
+		 * @param {*} summary Description of reason for undo to be stored as edit summary
```
Comment (vforge) [inline front/main/app/models/article-diff.js]: This JSDoc would be better: `@param {string|Array} [summary=[]]`

File: front/main/app/models/recent-wiki-activity.js [hunk 8]
```diff
@@ -1,11 +1,35 @@
...
+
 const defaultProps = 'user|userid|useravatar|parsedcomment|timestamp|title|ids|upvotes',
-	RecentWikiActivityModel = Ember.Object.extend({
-		init() {
-			this._super(...arguments);
-			this.recentChanges = {};
+	RecentWikiActivityModel = Ember.Object.extend(
+		revisionUpvotesMixin,
+		{
+			init() {
```
Comment (vforge) [inline front/main/app/models/recent-wiki-activity.js]: Missing JSDoc - here and in next function.

File: front/main/app/components/article-diff.js [hunk 9]
```diff
@@ -9,14 +9,11 @@ export default Ember.Component.extend(
...
-					'from_user',
-					this.get('currentUser.userId')
-				) : null;
-
-			return currentUserUpvote ? currentUserUpvote.id : null;
+		revisionUpvotes: Ember.inject.service(),
+		currentUserUpvoteId: Ember.computed('revisionUpvotes.upvotes.@each.count', 'currentUser.userId', function () {
+			const upvotes = this.get('revisionUpvotes.upvotes').findBy('revisionId', this.get('model.newId')) || [];
+
+			return upvotes.userUpvoteId;
```
Comment (bognix) [inline front/main/app/components/article-diff.js]: there might be a case when you are going to call userUpvoteId on an empty array. Even though it won't throw it looks odd

**Comment by bognix** [comment]
minor notes - LGTM

**Pull request merged**